### PR TITLE
Access Reviews | Expose group peer outlier insight and setting models

### PR DESCRIPTION
**Pull request by shubhamguptacal**
Schema update PR: 6248969

### CODE REVIEW

File: api-reference/beta/resources/membershipOutlierInsight.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
+---
+title: "membershipOutlierInsight resource type"
+description: "In the Azure AD access reviews, the membershipOutlierInsight resource represents insights provided to reviewers based on whether a user is a peer outlier w.r.t other group members"
```
Comment (FaithOmbongi) [inline api-reference/beta/resources/membershipOutlierInsight.md:3]:
```suggestion
description: "In the Azure AD access reviews, the membershipOutlierInsight resource represents insights provided to reviewers based on whether a user is a peer outlier in relation to other group members."
```

File: api-reference/beta/resources/grouppeeroutlierrecommendationinsightsettings.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+# groupPeerOutlierRecommendationInsightSettings resource type
+
+Namespace: microsoft.graph
+
+[!INCLUDE [beta-disclaimer](../../includes/beta-disclaimer.md)]
+
+[!INCLUDE [accessreviews-disclaimer-v2](../../includes/accessreviews-disclaimer-v2.md)]
+
+The **groupPeerOutlierRecommendationInsightSettings** allows you to configure group peer outlier insights to aid the reviewers to make decisions for an [accessReviewScheduleDefinition](accessreviewscheduledefinition.md) object.
```
Comment (FaithOmbongi) [inline api-reference/beta/resources/grouppeeroutlierrecommendationinsightsettings.md:18]:
```suggestion
The **groupPeerOutlierRecommendationInsightSettings** allows you to configure group peer outlier insights to aid reviewers to make decisions for an [accessReviewScheduleDefinition](accessreviewscheduledefinition.md) object.
```

File: api-reference/beta/resources/grouppeeroutlierrecommendationinsightsettings.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
+---
+title: "groupPeerOutlierRecommendationInsightSettings resource type"
+description: "In the Azure AD access reviews, the groupPeerOutlierRecommendationInsightSettings represents the settings associated with group peer outlier insights, and that is used to aid reviewers to make decisions."
```
Comment (FaithOmbongi) [inline api-reference/beta/resources/grouppeeroutlierrecommendationinsightsettings.md:3]:
```suggestion
description: "In the Azure AD access reviews, the groupPeerOutlierRecommendationInsightSettings represents the settings associated with group peer outlier insights, and that is used to aid reviewers to make decisions."
```

File: api-reference/beta/resources/grouppeeroutlierrecommendationinsightsettings.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+[!INCLUDE [accessreviews-disclaimer-v2](../../includes/accessreviews-disclaimer-v2.md)]
+
+The **groupPeerOutlierRecommendationInsightSettings** allows you to configure group peer outlier insights to aid the reviewers to make decisions for an [accessReviewScheduleDefinition](accessreviewscheduledefinition.md) object.
+
+Inherits from [accessReviewRecommendationInsightSetting](accessReviewRecommendationInsightSetting.md).
+
+## Properties
+| Property    | Type   | Description |
+| :---------------| :---------- | :---------- |
```
Comment (FaithOmbongi) [inline api-reference/beta/resources/grouppeeroutlierrecommendationinsightsettings.md:24]: Replace empty table with `None.`

File: api-reference/beta/resources/membershipOutlierInsight.md [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+# membershipOutlierInsight resource type
+
+Namespace: microsoft.graph
+
+[!INCLUDE [beta-disclaimer](../../includes/beta-disclaimer.md)]
+
+[!INCLUDE [accessreviews-disclaimer-v2](../../includes/accessreviews-disclaimer-v2.md)]
+
+Represents an insight provided to reviewers based on whether a user is a peer outlier w.r.t to other group members
```
Comment (FaithOmbongi) [inline api-reference/beta/resources/membershipOutlierInsight.md:18]:
```suggestion
Represents an insight provided to reviewers based on whether a user is a peer outlier in relation to other group members.
```

File: api-reference/beta/resources/membershipOutlierInsight.md [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+Inherits from [governanceInsight](governanceinsight.md).
+
+## Properties
+| Property    | Type   | Description |
+| :---------------| :---------- | :---------- |
+| memberId | String | Indicates the user Id of the member |
+| containerId | String | Indicates the id of the container(for ex: group Id) |
+| outlierContainerType | outlierContainerType | Indicates the type of container. For ex: group  |
+| outlierMemberType | outlierMemberType | Indicates the type of outlier member. For ex: user |
```
Comment (FaithOmbongi) [inline api-reference/beta/resources/membershipOutlierInsight.md:28]: Additional fixes:
+ You're missing the two properties inherited from governanceInsight.
+ Properties should be documented in alphabetical order

```suggestion
| memberId | String | Indicates the identifier of the user. |
| containerId | String | Indicates the identifier of the container, for example, a group ID. |
| outlierContainerType | outlierContainerType | Indicates the type of container. The possible values are: `group`, `unknownFutureValue`. |
| outlierMemberType | outlierMemberType | Indicates the type of outlier member. The possible values are: `user`, `unknownFutureValue`. |
```

File: api-reference/beta/resources/membershipOutlierInsight.md [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+}
+-->
+``` json
+{
+  "@odata.type": "#microsoft.graph.membershipOutlierInsight",
+  "memberId": "String",
+  "containerId": "String",
+  "outlierContainerType": "String",
+  "outlierMemberType": "String",
+}
```
Comment (FaithOmbongi) [inline api-reference/beta/resources/membershipOutlierInsight.md:48]: The stub generated JSON representation is complete as-is
```suggestion
The following is a JSON representation of the resource.
<!-- {
  "blockType": "resource",
  "keyProperty": "id",
  "@odata.type": "microsoft.graph.membershipOutlierInsight",
  "baseType": "microsoft.graph.governanceInsight",
  "openType": false
}
-->
``` json
{
  "@odata.type": "#microsoft.graph.membershipOutlierInsight",
  "id": "String (identifier)",
  "insightCreatedDateTime": "String (timestamp)",
  "memberId": "String",
  "containerId": "String",
  "outlierContainerType": "String",
  "outlierMemberType": "String"
}
```
```

File: changelog/Microsoft.AAD.Reporting.json [hunk 8]
```diff
@@ -1733,6 +1733,40 @@
...
       "SubArea": ""
+    },
+    {
+      "ChangeList": [
+        {
+          "Id": "b211329e-698e-4909-aa0c-4b967538bc13",
+          "ApiChange": "Resource",
+          "ChangedApiName": "membershipOutlierInsight",
+          "ChangeType": "Addition",
+          "Description": "Added the [membershipOutlierInsight](https://docs.microsoft.com/en-us/graph/api/resources/membershipOutlierInsight?view=graph-rest-beta) resource type. Use this type to configure insight settings based on the user's last sign-in date and time into a tenant or application.",
```
Comment (FaithOmbongi) [inline changelog/Microsoft.AAD.Reporting.json:1744]: Is this complex type really configurable? I thought it's read-only.

If confirmed to be read-only, just remove the second sentence from the changelog description.
```suggestion
          "Description": "Added the [membershipOutlierInsight](https://docs.microsoft.com/en-us/graph/api/resources/membershipOutlierInsight?view=graph-rest-beta) resource type. Use this type to configure insight settings based on the user's last sign-in date and time into a tenant or application.",
```

File: changelog/Microsoft.IdentityGovernance.AccessReviews.json [hunk 9]
```diff
@@ -1053,6 +1053,24 @@
...
       "SubArea": "Governance"
+    },
+    {
+      "ChangeList": [
+        {
+          "Id": "d8fdc071-6f2b-481e-ad87-29c4ff1145bf",
+          "ApiChange": "Resource",
+          "ChangedApiName": "groupPeerOutlierRecommendationInsightSettings",
+          "ChangeType": "Addition",
+          "Description": "Added the [groupPeerOutlierRecommendationInsightSettings](https://docs.microsoft.com/en-us/graph/api/resources/groupPeerOutlierRecommendationInsightSettings?view=graph-rest-beta) resource type. Use this type to configure insight settings based on the user's last sign-in date and time into a tenant or application.",
```
Comment (FaithOmbongi) [inline changelog/Microsoft.IdentityGovernance.AccessReviews.json:1064]:
```suggestion
          "Description": "Added the [groupPeerOutlierRecommendationInsightSettings](https://docs.microsoft.com/en-us/graph/api/resources/groupPeerOutlierRecommendationInsightSettings?view=graph-rest-beta) resource type. Use this type to configure access review insight settings based on the user's last sign-in date and time into a tenant or application.",
```

**Review by FaithOmbongi — CHANGES_REQUESTED** [review]
Thank you @shubhamguptacal for your work in this PR. I've left comments for a few fixes + the following:
- groupPeerOutlierRecommendationInsightSettings is the 2nd type that's derived from [accessReviewRecommendationInsightSetting](https://docs.microsoft.com/en-us/graph/api/resources/accessReviewRecommendationInsightSetting?view=graph-rest-beta). Add it to the list of derived types in the 1st paragraph of the accessReviewRecommendationInsightSetting doc - otherwise it's completely undiscoverable.
- membershipOutlierInsight is the 2nd type derived from [governanceInsight](https://docs.microsoft.com/en-us/graph/api/resources/governanceInsight?view=graph-rest-beta). Add it to the list of derived types in the 1st paragraph of the governanceInsight doc - otherwise, it's completely undiscoverable.
- Changelogs - assign the same ID to the changelog objects per file.


**Unclear linkage with access reviews.**
- One thing I don't think comes out clearly in this PR: The relationship between groupPeerOutlierRecommendationInsightSettings and membershipOutlierInsight. Could you clarify how they relate to each other? 
   - How do you configure the object in the accessReviewScheduleSettings object given it doesn't have or inherit any properties? 
   - groupPeerOutlierRecommendationInsightSettings object rolls up to the settings property while membershipOutlierInsight rolls up to the insights nav property in an accessReviewInstanceDecisionItem object. How do these two connect?

File: api-reference/beta/resources/membershipOutlierInsight.md [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+Namespace: microsoft.graph
+
+[!INCLUDE [beta-disclaimer](../../includes/beta-disclaimer.md)]
+
+[!INCLUDE [accessreviews-disclaimer-v2](../../includes/accessreviews-disclaimer-v2.md)]
+
+Represents an insight provided to reviewers based on whether a user is a peer outlier w.r.t to other group members
+
+Inherits from [governanceInsight](governanceinsight.md).
+
```
Comment (FaithOmbongi) [inline api-reference/beta/resources/membershipOutlierInsight.md:21]: This API highly likely exposes some methods. Please investigate and document them

Comment (shubhamguptacal) [reply api-reference/beta/resources/membershipOutlierInsight.md:21]: At a high level, there’s no CRUD exposed for all three insight models – governanceInsight,userLastSignInsight and memberhipOutlierInsight. The way these models surface up is through the insight navigation property exposed on AccessReviewInstanceDecisionItem resource type : [accessReviewInstanceDecisionItem resource type - Microsoft Graph beta | Microsoft Docs](https://docs.microsoft.com/en-us/graph/api/resources/accessreviewinstancedecisionitem?view=graph-rest-beta#relationships). Every accessReviewInstanceDecisionItem record has a collection of insights exposed through the “insights” navigation property. A developer can call the List Decisions API and pass in an expand clause for insights which will return a list of insights for all decision items. Here’s an example API call : 

GET /identityGovernance/accessReviews/definitions/{accessReviewScheduleDefinitionId}/instances/{accessReviewInstanceId}/decisions?$expand=insights
{
    "@odata.count": 1,
    "value": [
        {
            "id": "e6cafba0-cbf0-4748-8868-0810c7f4cc06",
            "accessReviewId": "6444d4fd-ab55-4608-8cf9-c6702d172bcc",
            "decision": "NotReviewed",
            "recommendation": "Approve",
            ....
            "insights": [
                {
                    "@odata.type": "#microsoft.graph.userSignInInsight",
                    "id": "00000000-0000-0000-0000-000000000000",
                    "insightCreatedDateTime": null,
                    "lastSignInDateTime": "2022-02-11T09:01:12Z"
                },
                {
                    "@odata.type": "#microsoft.graph.membershipOutlierInsight",
                    "id": "4e431ca4-df85-4d21-871f-77289959560e",
                    "insightCreatedDateTime": "2022-07-21T08:20:22Z",
                    "memberId": "53548fbc-bf77-48db-98e3-929a9c02197d",
                    "containerId": "98f41dad-68d5-42f6-a50f-ddd75c5c5539",
                    "outlierMemberType": "user",
                    "outlierContainerType": "group"
                }
            ]
        }
    ]
}

**Comment by shubhamguptacal** [comment]
> Thank you @shubhamguptacal for your work in this PR. I've left comments for a few fixes + the following:
> 
> * groupPeerOutlierRecommendationInsightSettings is the 2nd type that's derived from [accessReviewRecommendationInsightSetting](https://docs.microsoft.com/en-us/graph/api/resources/accessReviewRecommendationInsightSetting?view=graph-rest-beta). Add it to the list of derived types in the 1st paragraph of the accessReviewRecommendationInsightSetting doc - otherwise it's completely undiscoverable.
> * membershipOutlierInsight is the 2nd type derived from [governanceInsight](https://docs.microsoft.com/en-us/graph/api/resources/governanceInsight?view=graph-rest-beta). Add it to the list of derived types in the 1st paragraph of the governanceInsight doc - otherwise, it's completely undiscoverable.
> * membershipOutlierInsight needs two navigation properties - **container** and **member**. This also makes it seem like there are quite a few APIs that membershipOutlierInsight exposes?
>   
>   * If there are APIs exposed by this entity, let's also surface them in the TOC
> * Changelogs - assign the same ID to the changelog objects per file.
> 
> **Unclear linkage with access reviews.**
> 
> * One thing I don't think comes out clearly in this PR: The relationship between groupPeerOutlierRecommendationInsightSettings and membershipOutlierInsight. Could you clarify how they relate to each other?
>   
>   * How do you configure the object in the accessReviewScheduleSettings object given it doesn't have or inherit any properties?
>   * groupPeerOutlierRecommendationInsightSettings object rolls up to the settings property while membershipOutlierInsight rolls up to the insights nav property in an accessReviewInstanceDecisionItem object. How do these two connect?

At a high level, there are two functionalities that we’re exposing with this docs PR :
-	Configure peer outlier insights for access reviews recommendations – This can be done by using the “recommendationInsightSettings” property on the accessReviewScheduleDefinition object
-	Surface up peer outlier insights along with access review instance decision items – This is done by surfacing up the peerOutlierInsight model in the “insights” property on the AccessReviewInstanceDecisionItem object.


At a high level, there are different kinds of recommendation insights that a developer can enable for access review recommendations – userLastSignInRecommendationInsightSetting and groupPeerOutlierRecommendationInsightSetting. When creating a review, a developer can decide what kinds of insights they want to enable for that review. This is done by using the recommendationInsightSetting property. The recommendationInsightSetting property is a collection of insight settings. 

To enable userLastSignIn insight for recommendations, one can specify a payload like this : 

              "recommendationInsightSettings": [
            {
                "odata.type": "#microsoft.graph.userLastSignInRecommendationInsightSettings",
               "signInScope": "tenant",
               "recommendationLookBackDuration": "P30D"
            }
        ]

To enable both userLastSignIn and groupPeerOutlier insights, one can specify a payload like this : 

              "recommendationInsightSettings": [
            {
                "odata.type": "#microsoft.graph.userLastSignInRecommendationInsightSettings",
               "signInScope": "tenant",
               "recommendationLookBackDuration": "P30D"
            },
            {
                "odata.type": "#microsoft.graph.groupPeerOutlierRecommendationInsightSettings"
            }
        ]


At a high level, there’s no CRUD exposed for all three insight models – governanceInsight,userLastSignInsight and memberhipOutlierInsight. The way these models surface up is through the insight navigation property exposed on AccessReviewInstanceDecisionItem resource type : [accessReviewInstanceDecisionItem resource type - Microsoft Graph beta | Microsoft Docs](https://docs.microsoft.com/en-us/graph/api/resources/accessreviewinstancedecisionitem?view=graph-rest-beta#relationships). Every accessReviewInstanceDecisionItem record has a collection of insights exposed through the “insights” navigation property. A developer can call the List Decisions API and pass in an expand clause for insights which will return a list of insights for all decision items. Here’s an example API call : 

GET /identityGovernance/accessReviews/definitions/{accessReviewScheduleDefinitionId}/instances/{accessReviewInstanceId}/decisions?$expand=insights
{
    "@odata.count": 1,
    "value": [
        {
            "id": "e6cafba0-cbf0-4748-8868-0810c7f4cc06",
            "accessReviewId": "6444d4fd-ab55-4608-8cf9-c6702d172bcc",
            "decision": "NotReviewed",
            "recommendation": "Approve",
            ....
            "insights": [
                {
                    "@odata.type": "#microsoft.graph.userSignInInsight",
                    "id": "00000000-0000-0000-0000-000000000000",
                    "insightCreatedDateTime": null,
                    "lastSignInDateTime": "2022-02-11T09:01:12Z"
                },
                {
                    "@odata.type": "#microsoft.graph.membershipOutlierInsight",
                    "id": "4e431ca4-df85-4d21-871f-77289959560e",
                    "insightCreatedDateTime": "2022-07-21T08:20:22Z",
                    "memberId": "53548fbc-bf77-48db-98e3-929a9c02197d",
                    "containerId": "98f41dad-68d5-42f6-a50f-ddd75c5c5539",
                    "outlierMemberType": "user",
                    "outlierContainerType": "group"
                }
            ]
        }
    ]
}

**Review by FaithOmbongi — CHANGES_REQUESTED** [review]
Hi @shubhamguptacal - I've left suggestions for minor fixes.

**A suggestion:** Let us show customers how to configure and use these insights. 
+ You can add an example to the Create scheduleDefinition doc. Add an example 5 that has both multi-stage + recommendationInsightSettings (via stageSettings > recommendationInsightsSettings object) configured.
+ Also show an example for how to $expand the insights. You can replace [Example 2 in List decisions](https://docs.microsoft.com/en-us/graph/api/accessreviewinstance-list-decisions?view=graph-rest-beta&tabs=http#example-2-retrieve-all-decision-items-for-which-youre-a-reviewer-and-expand-the-definitions) with one that covers the existing scenario + expands the insights. Then you can explain what you see in the insights. I'd like customers to also see the container and members relationship.

**GAP:** 

I've gone back to the published API.md and I see this URL `https://graph.microsoft.com/beta/identityGovernance/insights/membershipOutlierInsights?$expand=member($select=id,givenName,jobTitle,mail,mobilePhone,preferredLanguage,surname,userPrincipalName)`. So it seems membershipOutlierInsights exposes at least a LIST and a GET operation to get tenant-wide insights. Please add the method docs for these operations. Let us show customers how to use these tenant-wide results.

File: api-reference/beta/resources/membershipOutlierInsight.md [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+## Properties
+| Property    | Type   | Description |
+| :---------------| :---------- | :---------- |
+| memberId | String | Indicates the identifier of the user. |
+| containerId | String | Indicates the identifier of the container, for example, a group ID. |
+| outlierContainerType | outlierContainerType | Indicates the type of container. The possible values are: `group`, `unknownFutureValue`. |
+| outlierMemberType | outlierMemberType | Indicates the type of outlier member. The possible values are: `user`, `unknownFutureValue`. |
```
Comment (FaithOmbongi) [inline api-reference/beta/resources/membershipOutlierInsight.md:28]: Let's indicate which ones support $filter. I've only found references for outlierContainerType and outlierMemberType . Pleaase fill the others if they support $filter.
```suggestion
| memberId | String | Indicates the identifier of the user. |
| containerId | String | Indicates the identifier of the container, for example, a group ID. |
| outlierContainerType | outlierContainerType | Indicates the type of container. The possible values are: `group`, `unknownFutureValue`. Supports `$filter` (`eq`). |
| outlierMemberType | outlierMemberType | Indicates the type of outlier member. The possible values are: `user`, `unknownFutureValue`. Supports `$filter` (`eq`). |
```

**Review by FaithOmbongi — CHANGES_REQUESTED** [review]

File: api-reference/beta/resources/membershipOutlierInsight.md (hunk 11 above)
Comment (shubhamguptacal) [reply api-reference/beta/resources/membershipOutlierInsight.md:28]: Currently $filter is not supported for any of these properties

**Comment by shubhamguptacal** [comment]
> Hi @shubhamguptacal - I've left suggestions for minor fixes.
> 
> **A suggestion:** Let us show customers how to configure and use these insights.
> 
> * You can add an example to the Create scheduleDefinition doc. Add an example 5 that has both multi-stage + recommendationInsightSettings (via stageSettings > recommendationInsightsSettings object) configured.
> * Also show an example for how to $expand the insights. You can replace [Example 2 in List decisions](https://docs.microsoft.com/en-us/graph/api/accessreviewinstance-list-decisions?view=graph-rest-beta&tabs=http#example-2-retrieve-all-decision-items-for-which-youre-a-reviewer-and-expand-the-definitions) with one that covers the existing scenario + expands the insights. Then you can explain what you see in the insights. I'd like customers to also see the container and members relationship.
> 
> **GAP:**
> 
> I've gone back to the published API.md and I see this URL `https://graph.microsoft.com/beta/identityGovernance/insights/membershipOutlierInsights?$expand=member($select=id,givenName,jobTitle,mail,mobilePhone,preferredLanguage,surname,userPrincipalName)`. So it seems membershipOutlierInsights exposes at least a LIST and a GET operation to get tenant-wide insights. Please add the method docs for these operations. Let us show customers how to use these tenant-wide results.

Hi @FaithOmbongi 

Thanks for your feedback! I've updated the PR based on your feedback. The gap that you listed is an API listed in the API.md file but it's not exposed through the schema at this time(Here's a PR for the schema exposed today : https://msazure.visualstudio.com/One/_git/AD-AggregatorService-Workloads/pullrequest/6248969). That's why that API is not documented in this PR

**Review by FaithOmbongi — APPROVED** [review]
Hi @shubhamguptacal  - Approving this PR as the actual content review is complete. I think the only pending thing was resolving the bug that affected the use of groupPeerOutlierRecommendationInsightSetting.

Feel free to push this PR to the merge queue is the bug has been resolved.

**Pull request merged**